src/test/java/com/mongodb/jdbc/MongoSQLMock.java

```diff
@@ -0,0 +1,193 @@
+import java.lang.reflect.Field;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.HashSet;
+import org.bson.BsonDocument;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.internal.util.reflection.FieldSetter;
+
+public abstract class MongoSQLMock {
+    static ConnectionString uri = new ConnectionString("mongodb://localhost:27017/admin");;
+    protected static String database = "test";
+    @Mock protected static MongoClient mongoClient;
+    @Mock protected static MongoDatabase mongoDatabase;
+    @Mock protected static AggregateIterable<MongoResultDoc> aggregateIterable;
+    @Mock protected static MongoCursor<MongoResultDoc> mongoCursor;
+
+    @InjectMocks
+    protected static MongoConnection mongoConnection = new MongoSQLConnection(uri, database);
+
+    private static Field getDeclaredFieldFromClassOrSuperClass(Class c, String fieldName)
+            throws NoSuchFieldException {
+        try {
+            return c.getDeclaredField(fieldName);
+        } catch (NoSuchFieldException e) {
+            Class superClass = c.getSuperclass();
+            if (superClass != null) {
+                return getDeclaredFieldFromClassOrSuperClass(superClass, fieldName);
+            }
+        }
+        throw new NoSuchFieldException(fieldName);
+    }
+
+    // reset the mock objects before every test case
+    protected static void resetMockObjs() throws NoSuchFieldException {
+        FieldSetter.setField(
+                mongoConnection,
+                getDeclaredFieldFromClassOrSuperClass(mongoConnection.getClass(), "mongoClient"),
+                mongoClient);
+        FieldSetter.setField(
+                mongoConnection,
+                getDeclaredFieldFromClassOrSuperClass(mongoConnection.getClass(), "isClosed"),
+                false);
+        FieldSetter.setField(
+                mongoConnection,
+                getDeclaredFieldFromClassOrSuperClass(mongoConnection.getClass(), "currentDB"),
+                database);
+
+        doNothing().when(mongoClient).close();
+        // Mock mongoDatabase
+        when(mongoConnection.getDatabase(anyString())).thenReturn(mongoDatabase);
+        when(mongoDatabase.withCodecRegistry(any())).thenReturn(mongoDatabase);
+        when(mongoDatabase.aggregate(any(), eq(MongoResultDoc.class)))
+                .thenReturn(aggregateIterable);
+        // Mock aggregateIterable
+        when(aggregateIterable.batchSize(anyInt())).thenReturn(aggregateIterable);
+        when(aggregateIterable.maxTime(anyLong(), any())).thenReturn(aggregateIterable);
+        when(aggregateIterable.cursor()).thenReturn(mongoCursor);
+        // Mock MongoCursor
+        when(mongoCursor.hasNext()).thenReturn(false);
+    }
+
+    // to replace lambda as input in the testExceptionAfterConnectionClosed
+    interface TestInterface {
+        void test() throws SQLException;
+    }
+
+    protected static MongoJsonSchema generateMongoJsonSchema() {
```

Comment:
Note that both the datasources and keys within them are not in lexicographical order to show that they are properly sorted in the test